lib/utils/genmap/genmap_test.go

```diff
@@ -0,0 +1,183 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package genmap
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"golang.org/x/sync/errgroup"
+)
+
+// TestCaching verifies the basic expected behavior of foreground operations.
+func TestCaching(t *testing.T) {
+	t.Parallel()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// set up a genmap with a long regen interval
+	var counter int
+	gm := New(Config[string, int]{
+		RegenInterval: time.Hour,
+		Generator: func(_ context.Context, key string) (int, error) {
+			counter++
+			return counter, nil
+		},
+	})
+	defer gm.Close()
+
+	// verify that many concurrent calls result in only a single call to
+	// the underlying generator.
+	var eg errgroup.Group
+	for i := 0; i < 100; i++ {
+		eg.Go(func() error {
+			n, err := gm.Get(ctx, "some-key")
+			if err != nil {
+				return err
+			}
+
+			if n != 1 {
+				return fmt.Errorf("expected 1, got %d", n)
+			}
+
+			return nil
+		})
+	}
+
+	require.NoError(t, eg.Wait())
+
+	// force an early regen
+	gm.Generate("some-key")
+
+	// verify that regen occurs
+	require.Eventually(t, func() bool {
+		n, _ := gm.Get(ctx, "some-key")
+		return n == 2
+	}, time.Second*30, time.Millisecond*100)
+}
+
+// TestConcurrentTermination verifies that concurrently terminating background ops does not
+// interfere with the ability of each individual Get to yield a sensible/expected value.
+func TestConcurrentTermination(t *testing.T) {
+	t.Parallel()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// set up a genmap with a short regen interval
+	var counter int
+	gm := New(Config[string, int]{
+		RegenInterval: time.Hour,
+		Generator: func(_ context.Context, key string) (int, error) {
+			counter++
+			return counter, nil
+		},
+	})
+	defer gm.Close()
+
+	var eg errgroup.Group
+	for i := 0; i < 200; i++ {
+		if i%2 == 0 {
+			eg.Go(func() error {
+				for j := 0; j < 100; j++ {
+					n, err := gm.Get(ctx, "some-key")
+					if err != nil {
+						return err
+					}
+					if n == 0 {
+						return fmt.Errorf("expected nonzero n")
+					}
+				}
+				return nil
+			})
+		} else {
+			eg.Go(func() error {
+				gm.Terminate("some-key")
+				return nil
+			})
+		}
+	}
+
+	require.NoError(t, eg.Wait())
+
+	// sanity check to assert that concurrent termination really happened (in practice we expect
+	// this value to always be *much* larger than 2).
```

Comment:
in worst case scenario though, if all `gm.Terminate` happen before all `gm.Get`, the value would be 1?

Reply:
Correct.  Though the odds of that should be are astronomically small, as that would require all 100 sequential calls to `Get` in all 100 of the getter routines to complete before any of the 100 calls to `Terminate` complete.

Reply:
> Though the odds of that should be are astronomically small

Agreed. I guess just my paranoia, maybe do a `gm.Get` before the loop just to make sure. Up to you.